### DISCUSSION TITLE
TC-Malloc Library Support

### DISCUSSION THREAD

**Discussion by blue-j** [ImageMagick/ImageMagick#5179] [Help] [answered]
I successfully installed tcmalloc and passed all its tests.  Yet when I try to build `--with-tcmalloc=yes` it fails.

Scrolling through the build comments I find this:

```
checking tcmalloc library support... 
checking for mallinfo in -ltcmalloc_minimal... no
checking if tcmalloc memory allocation library is complete... no
```

This is all new to me!  Does anyone know how I might be able to remedy this and get IM to build with tcmalloc?

Thanks in advance for your consideration!

-J

**Comment by blue-j** [comment]
Just in case this is helpful, this is what I did, all at `/usr/src` (Ubuntu 20):

```
sudo apt install apt-transport-https curl gnupg
curl -fsSL https://bazel.build/bazel-release.pub.gpg | gpg --dearmor > bazel.gpg
sudo mv bazel.gpg /etc/apt/trusted.gpg.d/
echo "deb [arch=amd64] https://storage.googleapis.com/bazel-apt stable jdk1.8" | sudo tee /etc/apt/sources.list.d/bazel.list
sudo apt update && sudo apt install bazel
sudo mkdir Source; cd Source
sudo git clone https://github.com/google/tcmalloc.git
cd tcmalloc
sudo bazel test //tcmalloc/...
sudo wget https://www.imagemagick.org/download/ImageMagick.tar.gz
sudo tar xvzf ImageMagick.tar.gz
cd ImageMagick-7*
sudo ./configure --with-tcmalloc=yes --with-lqr=yes --with-bzlib=yes --with-djvu=yes --with-fftw=yes --with-fontconfig=yes --with-freetype=yes --with-heic=yes --with-jbig=yes --with-jpeg=yes --with-jxl=yes --with-lcms=yes --with-openexr=yes --with-pango=yes  --with-png=yes --with-rsvg=yes --with-tiff=yes --with-webp=yes --with-wmf=yes --with-xml=yes --with-zlib=yes
sudo make
sudo make install
sudo ldconfig
```

**Reply by blue-j** [reply]
==============================================================================
ImageMagick 7.1.0-35 is configured as follows. Please verify that this
configuration matches your expectations.

Host system type: x86_64-pc-linux-gnu
Build system type: x86_64-pc-linux-gnu

                  Option                        Value
------------------------------------------------------------------------------
Shared libraries  --enable-shared=yes		yes
Static libraries  --enable-static=yes		yes
Build utilities   --with-utilities=yes		yes
Module support    --with-modules=no		no
GNU ld            --with-gnu-ld=yes		yes
Quantum depth     --with-quantum-depth=16	16
High Dynamic Range Imagery
                  --enable-hdri=yes		yes

Install documentation:				yes

Memory allocation library:
  JEMalloc          --with-jemalloc=no		no
  TCMalloc          --with-tcmalloc=yes		no
  UMem              --with-umem=no		no

Delegate library configuration:
  BZLIB             --with-bzlib=yes		yes
  Autotrace         --with-autotrace=no		no
  DJVU              --with-djvu=yes		yes
  DPS               --with-dps=yes		no
  FFTW              --with-fftw=yes		yes
  FLIF              --with-flif=no		no
  FlashPIX          --with-fpx=yes		no
  FontConfig        --with-fontconfig=yes	yes
  FreeType          --with-freetype=yes		yes
  Ghostscript lib   --with-gslib=no		no
  Graphviz          --with-gvc=yes		no
  HEIC              --with-heic=yes		yes
  JBIG              --with-jbig=yes		yes
  JPEG v1           --with-jpeg=yes		yes
  JPEG XL           --with-jxl=yes		yes
  LCMS              --with-lcms=yes		yes
  LQR               --with-lqr=yes		yes
  LTDL              --with-ltdl=no		no
  LZMA              --with-lzma=yes		yes
  Magick++          --with-magick-plus-plus=yes	yes
  OpenEXR           --with-openexr=yes		yes
  OpenJP2           --with-openjp2=yes		yes
  PANGO             --with-pango=yes		yes
  PERL              --with-perl=no		no
  PNG               --with-png=yes		yes
  RAQM              --with-raqm=yes		no
  RAW               --with-raw=yes		no
  RSVG              --with-rsvg=yes		yes
  TIFF              --with-tiff=yes		yes
  WEBP              --with-webp=yes		yes
  WMF               --with-wmf=yes		yes
  X11               --with-x=			yes
  XML               --with-xml=yes		yes
  ZIP               --with-zip=yes		no
  ZLIB              --with-zlib=yes		yes
  ZSTD              --with-zstd=yes		yes

Delegate program configuration:
  GhostPCL          None			pcl6 (unknown)
  GhostXPS          None			gxps (unknown)
  Ghostscript       None			gs (9.50)

Font configuration:
  Apple fonts       --with-apple-font-dir=default	
  Dejavu fonts      --with-dejavu-font-dir=default	none
  Ghostscript fonts --with-gs-font-dir=default		/usr/share/fonts/type1/gsfonts/
  URW-base35 fonts  --with-urw-base35-font-dir=default	none
  Windows fonts     --with-windows-font-dir=default	none

X11 configuration:
  X_CFLAGS        = 
  X_PRE_LIBS      =  -lSM -lICE
  X_LIBS          = 
  X_EXTRA_LIBS    = 
######
Options used to compile and link:
  PREFIX          = /usr/local
  EXEC-PREFIX     = /usr/local
  VERSION         = 7.1.0-35
  CC              = gcc
  CFLAGS          = -I/usr/include/libxml2   -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -pthread -I/usr/include/librsvg-2.0 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libpng16  -I/usr/include/pango-1.0 -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/cairo -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/include/pixman-1 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/OpenEXR  -I/usr/include/lqr-1 -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include -I/usr/local/include/openjpeg-2.5  -I/usr/local/include -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/uuid -I/usr/include/freetype2 -I/usr/include/libpng16 -pthread    -fopenmp -Wall -g -O2 -mtune=haswell -fexceptions -pthread -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16
  CPPFLAGS        =  -DMAGICKCORE_HDRI_ENABLE=1 -DMAGICKCORE_QUANTUM_DEPTH=16 
  PCFLAGS         = 
  DEFS            = -DHAVE_CONFIG_H
  LDFLAGS         =  
  LIBS            =  -ljbig -llcms2 -ltiff -lfreetype  -ljpeg -L/usr/local/lib -ljxl -ljxl_threads  -llqr-1 -lglib-2.0 -lpng16 -lz  -ldjvulibre -lfftw3   -lfontconfig -lfreetype -L/usr/local/lib -lheif -lwebpmux -lwebpdemux -lwebp -lwebp -lwmflite  -lXext -lXt   -lSM -lICE -lX11  -llzma -lbz2 -lIlmImf -lImath -lHalf -lIex -lIexMath -lIlmThread -lpthread -L/usr/local/lib -lopenjp2 -lpangocairo-1.0 -lpango-1.0 -lgobject-2.0 -lglib-2.0 -lharfbuzz -lcairo  -lrsvg-2 -lm -lgio-2.0 -lgdk_pixbuf-2.0 -lgobject-2.0 -lglib-2.0 -lcairo -lxml2  -lz     -lm    -lpthread 
  CXX             = g++
  CXXFLAGS        =  -pthread
  FEATURES        = DPC HDRI Cipher OpenMP
  DELEGATES       = bzlib djvu fftw fontconfig freetype heic jbig jng jpeg jxl lcms lqr lzma openexr openjp2 pango png ps rsvg tiff webp wmf x xml zlib zstd
######
==============================================================================

**Comment by dlemstra** [comment, marked as answer]
I don't see a step here that looks like it will install `tcmalloc` on your machine?
```
sudo git clone https://github.com/google/tcmalloc.git
cd tcmalloc
sudo bazel test //tcmalloc/...
```

**Reply by blue-j** [reply]
My understanding from here is that the bazel test command tests AND installs. That's what it looked like when I did it as well. 

[https://google.github.io/tcmalloc/quickstart.html]

"Congratulations! You’ve installed TCMalloc"

I am new to Bazel and apologize in advance if I've misunderstood anything. I followed the instructions on the link to the T. But maybe I missed something?

**Reply by dlemstra** [reply]
I also have no clue how bazel works but running tests did not look like it would install a library. Do you see a `libtcmalloc` file in `/usr/lib` after running those tests?

**Reply by dlemstra** [reply]
Maybe you could use `libgoogle-perftools-dev` instead of building from source? Haven't tried if this work but that package will install `/usr/lib/x86_64-linux-gnu/libtcmalloc.a`.

**Reply by blue-j** [reply]
Thank you for this suggestion!  I was able to easily install that package and see libtcmalloc.a at that path.  Will ImageMagick be able to find that, or do I need to move it?  Thanks, Dirk!  I've spent about 7 hours on this without joy.   - J

**Reply by blue-j** [reply]
It worked!!  Woohoo!!!  You are a god.

**Reply by dlemstra** [reply]
I did not check our `configure.ac` file before posting this but it seems that we look for `tcmalloc_minimal` . But that also seems to be installed with that package. I did not test yet if ImageMagick could find that library but I was hoping you could quickly test that yourself.

*edit* I am happy to hear that this worked for you.

**Reply by blue-j** [reply]
i'm surely being daft, but how do i mark this as answered?  the only option I see is to mark the wrong post as the answer.
